### PR TITLE
FastTeamChainLoader - implementation and happy-path tests

### PR DESCRIPTION
**Pull request by maxtaco**
- the FastTeamChainLoader takes many shortcuts when loading a team, and is only willing to output team keys and the team's name
- this PR has the full loader, minus the "audit" phase, which is left for a future PR.
- there are tests for happy-paths of the major cases I think of, but not corrupt-path tests just yet
- also, it isn't plugged into anything other than unit tests.

### CODE REVIEW

File: go/teams/loader_ctx.go [hunk 1]
```diff
@@ -44,6 +45,62 @@ func NewLoaderContextFromG(g *libkb.GlobalContext) LoaderContext {
...
+	LegacyTLFUpgrade []keybase1.TeamGetLegacyTLFUpgrade `json:"legacy_tlf_upgrade"`
+}
+
+func (r *rawTeam) GetAppStatus() *libkb.AppStatus {
+	return &r.Status
+}
+
+func (t *rawTeam) unpackLinks(ctx context.Context) ([]*chainLinkUnpacked, error) {
+	if t == nil {
+		return nil, nil
```
Comment (mlsteele) [inline go/teams/loader_ctx.go]: This seems bad for a receiver. Caller should not call methods on nil objects. I consider it a go design bug that this is possible

File: go/teams/fast.go [hunk 2]
```diff
@@ -72,43 +95,945 @@ func (f *FastTeamChainLoader) verifyTeamNameViaParentLoad(m libkb.MetaContext, i
...
+	if err != nil {
+		return nil, err
+	}
+	f.updateCache(m, state)
+
+	return f.toResult(m, arg, state)
+}
+
+// newLRU installs a new LRU for the loader and purges the old one. Does a swap to avoid race conditions
+// around logging out.
```
Comment (mlsteele) [inline go/teams/fast.go]: Why swap? The lru is threadsafe. Wouldn't it be simpler to initialize the lru in the constructor, Purge here, and do away with the `lruMutex`?

File: go/teams/fast.go [hunk 3]
```diff
@@ -72,43 +95,945 @@ func (f *FastTeamChainLoader) verifyTeamNameViaParentLoad(m libkb.MetaContext, i
...
+		}
+		prev = link.LinkTriple()
+	}
+	return nil
+}
+
+// audit runs probabilistic merkle tree audit on the new links, to make sure that the server isn't
+// running odd-even-style attacks against members in a group.
+func (f *FastTeamChainLoader) audit(m libkb.MetaContext, id keybase1.TeamID, isPublic bool, newLinks []*chainLinkUnpacked) (err error) {
+	return nil
```
Comment (mlsteele) [inline go/teams/fast.go]: put todo maybe ticket number

File: go/teams/fast.go (hunk 2 above)
Comment (maxtaco) [reply go/teams/fast.go]: what about: (1) start load; (2) user logs out and logs back in; (3) finish load.  now user B has user A's cache entry.

Comment (mlsteele) [reply go/teams/fast.go]: How do the strategies differ when it comes to that problem?

Comment (maxtaco) [reply go/teams/fast.go]: in my strategy, you're putting the cache to the old LRU, which is going to be garbage collected since nothing points to it.

Comment (mlsteele) [reply go/teams/fast.go]: `updateCache` doesn't take the old LRU as a parameter, it calls `getLRU()`. If the LRU was only `getLRU()`d once per Load call then I'd see what you're saying. Then that race would still return the team which is bad, but not store it which is good.

**Comment by maxtaco** [comment]
@mlsteele do you know how i might have broken this?

```2018-08-16 15:44:02.48475 timebuckets.go:48      : [D] TimeBucket loadKeyCache.loadKeyV2 [time=0s] [tags:LT=9Yqb2an-4FcO]
2018-08-16 15:44:02.48477 globals.go:1122        : [I] + G#GetPerUserKeyring
2018-08-16 15:44:02.48479 loader_ctx.go:201      : [I] - G#GetPerUserKeyring -> ERROR: PerUserKeyring unavailable with no UID
2018-08-16 15:44:02.48481 loader.go:246          : [D] - TeamLoader#load2(2123209b98b16083c69c91152b861724, public:false) -> loading team secrets: PerUserKeyring unavailable with no UID [time=6.024003ms] [tags:LT=9Yqb2an-4FcO]
	loader_chain_test.go:184: got error: [*errors.errorString] loading team secrets: PerUserKeyring unavailable with no UID
```

File: go/teams/fast.go (hunk 2 above)
Comment (maxtaco) [reply go/teams/fast.go]: ok, i see.  i can fix that.

Comment (maxtaco) [reply go/teams/fast.go]: ok, fixed

**Comment by maxtaco** [comment]
@mlsteele am I calling into the non-mock by accident, and I should be calling into the mock? Can't see how i did that though.

File: go/teams/loader2.go [hunk 4]
```diff
@@ -824,15 +824,20 @@ func (l *TeamLoader) checkReaderKeyMaskCoverage(ctx context.Context,
...
+}
+
+func unboxPerTeamSecrets(m libkb.MetaContext, box *TeamBox, prevs map[keybase1.PerTeamKeyGeneration]prevKeySealedEncoded) (keybase1.PerTeamKeyGeneration, []keybase1.PerTeamKeySeed, error) {
+
 	if box == nil {
 		return 0, nil, fmt.Errorf("no key box from server")
 	}
 
-	userKey, err := l.perUserEncryptionKey(ctx, box.PerUserKeySeqno)
+	userKey, err := perUserEncryptionKey(m, box.PerUserKeySeqno)
```
Comment (mlsteele) [inline go/teams/loader2.go]: It's here

Comment (mlsteele) [reply go/teams/loader2.go]: mmaybe not

Comment (maxtaco) [reply go/teams/loader2.go]: i think so too. ok, thanks

Comment (maxtaco) [reply go/teams/loader2.go]: Worked, thank you!

File: go/teams/fast.go [hunk 5]
```diff
@@ -72,43 +95,946 @@ func (f *FastTeamChainLoader) verifyTeamNameViaParentLoad(m libkb.MetaContext, i
...
+	if s != nil {
+		tmp := s.DeepCopy()
+		return &tmp
+	}
+	return &keybase1.FastTeamData{
+		PerTeamKeySeedsUnverified: make(map[keybase1.PerTeamKeyGeneration]keybase1.PerTeamKeySeed),
+		ReaderKeyMasks:            make(map[keybase1.TeamApplication](map[keybase1.PerTeamKeyGeneration]keybase1.MaskB64)),
+		Chain: keybase1.FastTeamSigChainState{
+			ID:                      arg.ID,
+			Public:                  arg.Public,
```
[the remark this answers is not in the document]
Comment (maxtaco) [reply go/teams/fast.go]: but i'm using this param to actually load the team?

[the remark this answers is not in the document]
Comment (mlsteele) [reply go/teams/fast.go]: yeah ok

[the remark this answers is not in the document]
Comment (maxtaco) [reply go/teams/fast.go]: i'll assert for now that the two agree when entering the whole thing.

**Comment by maxtaco** [comment]
This is a great idea about the fallback.  Let's plumb an error through and then the next ticket can pick that up.

**Pull request merged**